design/multiple-listeners-design.md

```diff
@@ -0,0 +1,149 @@
+# Support for multiple Listeners
+
+Status: Draft
+
+## Abstract
+This document proposes changes to Contour in order to support an arbitrary number of Listeners.
+
+## Background
+Historically, Contour has always supported configuring two Envoy Listeners: one for HTTP, and one for HTTPS.
+Root HTTPProxies without TLS enabled have been configured on the HTTP Listener, and those with TLS enabled have been configured on the HTTPS Listener.
+Now that Contour implements Gateway API, it's necessary to support any number of Listeners in order to be fully conformant with the Gateway API spec.
+
+It's important to distinguish between **Envoy Listeners** and **Gateway Listeners**, since they are related but not the same.
+
+An **Envoy Listener** can be thought of simply as Envoy listening on a given port.
+
+A **Gateway Listener** encompasses listening on a certain "exposed" port (typically defined on an Envoy LoadBalancer service), which is mapped to an underlying Envoy Listener, for requests with a given hostname, and optionally with certain TLS details.
+In Envoy terminology, this more closely corresponds to the combination of a Listener, a FilterChain, and a RouteConfiguration virtual host.
+
+## Goals
+- Support Gateways, and associated HTTPRoutes/TLSRoutes, with Listeners configured on >2 ports.
+- Don't break existing functionality for HTTPProxy/Ingress.
+
+## Non Goals
+- Support for HTTPProxy/Ingress attaching to additional Listeners.
+- Support for new route types, e.g. TCPRoute.
+
+## High-Level Design
+Broadly speaking, Contour can operate in one of two modes, which we will call **Gateway mode** and **non-Gateway mode**.
+In Gateway mode, Contour is configured to process a Gateway API Gateway, and gets its Listener configuration from there.
+Non-Gateway mode is the traditional mode of operation, where Listener configuration comes from CLI flags, a config file, and/or a ContourConfiguration resource, plus the Envoy service YAML, and HTTPProxy and/or Ingress is used for routing configuration.
+
+For non-Gateway mode, this design should be fully backwards-compatible.
+There should be no changes observed by a user of Contour in non-Gateway mode.
+
+For Gateway mode, the following changes are needed:
```

Comment:
maybe add a little comment on "Gateway mode" which I assume is when you're using the Gateway provisioner etc., vs. "non-Gateway" which is when you're not (using "static" Gateway mode etc.)

(sidebar, maybe some compatibility notes to add about "static" mode?)

Reply:
> maybe add a little comment on "Gateway mode" which I assume is when you're using the Gateway provisioner etc., vs. "non-Gateway" which is when you're not (using "static" Gateway mode etc.)

Are you looking for something more than L29-31? 

> (sidebar, maybe some compatibility notes to add about "static" mode?)

Yep good callout, the tricky thing here is that users will have to update the Envoy service to expose the correct ports, which means they need to understand the port mapping Contour will use.